In [5]:
import pandas as pd
from sqlalchemy import create_engine

In [6]:
df=pd.read_csv("indian_techjobs_2026.csv")

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23201 entries, 0 to 23200
Data columns (total 32 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   job_id               23201 non-null  int64  
 1   job_title            23201 non-null  object 
 2   company_name         23201 non-null  object 
 3   company_rating       23201 non-null  float64
 4   location             23201 non-null  object 
 5   scraped_city         23201 non-null  object 
 6   role_category        23201 non-null  object 
 7   experience_raw       23201 non-null  object 
 8   experience_min_yrs   23201 non-null  int64  
 9   experience_max_yrs   23201 non-null  int64  
 10  salary_raw           23201 non-null  object 
 11  salary_min_lpa       23201 non-null  float64
 12  salary_max_lpa       23201 non-null  float64
 13  salary_disclosed     23201 non-null  bool   
 14  skills_required      23201 non-null  object 
 15  skills_count         23201 non-null 

In [11]:
df.shape

(23201, 32)

In [23]:
df.columns

Index(['job_id', 'job_title', 'company_name', 'company_rating', 'location',
       'scraped_city', 'role_category', 'experience_raw', 'experience_min_yrs',
       'experience_max_yrs', 'salary_raw', 'salary_min_lpa', 'salary_max_lpa',
       'salary_disclosed', 'skills_required', 'skills_count',
       'job_description', 'posted_date_raw', 'work_mode',
       'company_size_bucket', 'job_url', 'data_source', 'scraped_at',
       'salary_tier', 'experience_tier', 'is_senior', 'primary_city',
       'skill_domain', 'salary_midpoint_lpa', 'days_since_posted',
       'is_fresher_friendly', 'salary_negotiable'],
      dtype='object')

In [13]:
df.describe()

,job_id,company_rating,experience_min_yrs,experience_max_yrs,salary_min_lpa,salary_max_lpa,skills_count,salary_midpoint_lpa,days_since_posted
count,23201.000000,23201.000000,23201.000000,23201.000000,23201.000000,23201.000000,23201.000000,23201.000000,23201.000000
mean,11601.000000,3.586449,4.135080,7.829059,1.383952,2.271790,7.604198,1.827872,15.698935
std,6697.696134,0.514250,2.933976,3.946834,4.841570,7.469383,1.456530,6.109373,7.575881
min,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,5801.000000,3.400000,2.000000,5.000000,0.000000,0.000000,8.000000,0.000000,7.000000
50%,11601.000000,3.600000,4.000000,8.000000,0.000000,0.000000,8.000000,0.000000,21.000000
75%,17401.000000,3.800000,6.000000,10.000000,0.000000,0.000000,8.000000,0.000000,21.000000
max,23201.000000,5.000000,25.000000,31.000000,85.000000,90.000000,8.000000,87.500000,90.000000


In [17]:
df.isnull().sum()

job_id                  0
job_title               0
company_name            0
company_rating          0
location                0
scraped_city            0
role_category           0
experience_raw          0
experience_min_yrs      0
experience_max_yrs      0
salary_raw              0
salary_min_lpa          0
salary_max_lpa          0
salary_disclosed        0
skills_required         0
skills_count            0
job_description         0
posted_date_raw         0
work_mode               0
company_size_bucket     0
job_url                53
data_source             0
scraped_at              0
salary_tier             0
experience_tier         0
is_senior               0
primary_city            0
skill_domain            0
salary_midpoint_lpa     0
days_since_posted       0
is_fresher_friendly     0
salary_negotiable       0
dtype: int64

In [25]:
df.duplicated().sum()

0

# ====================================
# DATA PROFILING
# ====================================

In [28]:
print("Rows, Columns:", df.shape)
print("\nColumn Names:")
print(df.columns.tolist())

Rows, Columns: (23201, 32)

Column Names:
['job_id', 'job_title', 'company_name', 'company_rating', 'location', 'scraped_city', 'role_category', 'experience_raw', 'experience_min_yrs', 'experience_max_yrs', 'salary_raw', 'salary_min_lpa', 'salary_max_lpa', 'salary_disclosed', 'skills_required', 'skills_count', 'job_description', 'posted_date_raw', 'work_mode', 'company_size_bucket', 'job_url', 'data_source', 'scraped_at', 'salary_tier', 'experience_tier', 'is_senior', 'primary_city', 'skill_domain', 'salary_midpoint_lpa', 'days_since_posted', 'is_fresher_friendly', 'salary_negotiable']


In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23201 entries, 0 to 23200
Data columns (total 32 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   job_id               23201 non-null  int64  
 1   job_title            23201 non-null  object 
 2   company_name         23201 non-null  object 
 3   company_rating       23201 non-null  float64
 4   location             23201 non-null  object 
 5   scraped_city         23201 non-null  object 
 6   role_category        23201 non-null  object 
 7   experience_raw       23201 non-null  object 
 8   experience_min_yrs   23201 non-null  int64  
 9   experience_max_yrs   23201 non-null  int64  
 10  salary_raw           23201 non-null  object 
 11  salary_min_lpa       23201 non-null  float64
 12  salary_max_lpa       23201 non-null  float64
 13  salary_disclosed     23201 non-null  bool   
 14  skills_required      23201 non-null  object 
 15  skills_count         23201 non-null 

In [32]:
missing = df.isnull().sum().sort_values(ascending=False)
missing[missing > 0]

job_url    53
dtype: int64

In [34]:
print("Duplicate Rows:", df.duplicated().sum())

Duplicate Rows: 0


In [36]:
df.describe()

,job_id,company_rating,experience_min_yrs,experience_max_yrs,salary_min_lpa,salary_max_lpa,skills_count,salary_midpoint_lpa,days_since_posted
count,23201.000000,23201.000000,23201.000000,23201.000000,23201.000000,23201.000000,23201.000000,23201.000000,23201.000000
mean,11601.000000,3.586449,4.135080,7.829059,1.383952,2.271790,7.604198,1.827872,15.698935
std,6697.696134,0.514250,2.933976,3.946834,4.841570,7.469383,1.456530,6.109373,7.575881
min,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,5801.000000,3.400000,2.000000,5.000000,0.000000,0.000000,8.000000,0.000000,7.000000
50%,11601.000000,3.600000,4.000000,8.000000,0.000000,0.000000,8.000000,0.000000,21.000000
75%,17401.000000,3.800000,6.000000,10.000000,0.000000,0.000000,8.000000,0.000000,21.000000
max,23201.000000,5.000000,25.000000,31.000000,85.000000,90.000000,8.000000,87.500000,90.000000


In [38]:
df.sample(5, random_state=42)

,job_id,job_title,company_name,company_rating,location,scraped_city,role_category,experience_raw,experience_min_yrs,experience_max_yrs,...,scraped_at,salary_tier,experience_tier,is_senior,primary_city,skill_domain,salary_midpoint_lpa,days_since_posted,is_fresher_friendly,salary_negotiable
259,260,Data Scientist,Gallagher,3.6,Bengaluru,Bangalore,Data Scientist,3-8 Yrs,3,8,...,10-06-2025,Undisclosed,Mid (3-5 Yrs),False,Bangalore,AI/ML/DL,0.0,21,False,False
21183,21184,Deputy Manager,Srmb Srijan,3.5,Kolkata,Kolkata,Business Analyst,8-10 Yrs,8,10,...,10-06-2025,Undisclosed,Senior (6-8 Yrs),True,Kolkata,Business Intelligence,0.0,21,False,False
20313,20314,Sap Po Consultant,Tata Consultancy Services,3.3,Noida,Noida,Business Analyst,5-8 Yrs,5,8,...,10-06-2025,Undisclosed,Mid (3-5 Yrs),False,Noida,Business Intelligence,0.0,2,False,False
8541,8542,IoT Data Analysis Professional,Segula Technologies India,3.4,Chennai,Chennai,Data Analyst,5-10 Yrs,5,10,...,10-06-2025,Undisclosed,Mid (3-5 Yrs),False,Chennai,AI/ML/DL,0.0,1,False,False
2004,2005,Azure Data Engineer - ADF & PySpark (PAN India),Tata Consultancy Services,3.3,"Pune, Hyderabad, Bengaluru",Pune,Data Scientist,6-10 Yrs,6,10,...,10-06-2025,Undisclosed,Senior (6-8 Yrs),False,Pune,Cloud & DevOps,0.0,7,False,False


# First Business Questions

## 1. Which cities have the most job postings?

In [7]:
df['scraped_city'].value_counts()

scraped_city
Bangalore    3685
Mumbai       3368
Chennai      2693
Pune         2557
Delhi        2176
Noida        2147
Remote       1753
Kolkata      1701
Gurgaon      1636
Ahmedabad    1405
Hyderabad      80
Name: count, dtype: int64

## 2. Which companies are hiring the most?

In [8]:
df['company_name'].value_counts().head(20)

company_name
Tata Consultancy Services    653
Accenture                    438
EY                           340
Leading Client               243
Infosys                      209
Capgemini                    195
Photon                       193
Logic Planet                 185
Hexaware Technologies        147
Sparix Global                126
Paytm                        111
Fractal Analytics            108
Krazy Mantra It              102
Han Digital Solution          98
Crisil                        98
Diverse Lynx                  94
Iris Software                 91
Nagarro                       90
Cognizant                     88
Tech Stalwart Solution        86
Name: count, dtype: int64

## 3. Which job roles are most common?

In [9]:
df['role_category'].value_counts()

role_category
Data Scientist               6455
Data Analyst                 4729
Business Analyst             4505
Machine Learning Engineer    4004
Data Engineer                1922
Python Developer             1586
Name: count, dtype: int64

## 4. Remote vs Hybrid vs Onsite

In [10]:
df['work_mode'].value_counts()

work_mode
On-site    18701
Hybrid      2347
Remote      2153
Name: count, dtype: int64

## 5. Fresher-friendly jobs

In [11]:
df['is_fresher_friendly'].value_counts()

is_fresher_friendly
False    19019
True      4182
Name: count, dtype: int64

## 6. Salary Statistics

In [12]:
df[['salary_min_lpa',
    'salary_max_lpa',
    'salary_midpoint_lpa']].describe()

,salary_min_lpa,salary_max_lpa,salary_midpoint_lpa
count,23201.000000,23201.000000,23201.000000
mean,1.383952,2.271790,1.827872
std,4.841570,7.469383,6.109373
min,0.000000,0.000000,0.000000
25%,0.000000,0.000000,0.000000
50%,0.000000,0.000000,0.000000
75%,0.000000,0.000000,0.000000
max,85.000000,90.000000,87.500000


## 7. Experience Distribution

In [13]:
df['experience_tier'].value_counts()

experience_tier
Mid (3-5 Yrs)              10112
Junior (0-2 Yrs)            4915
Senior (6-8 Yrs)            4810
Fresher                     2219
Lead/Architect (9+ Yrs)     1145
Name: count, dtype: int64

## 8. Skill Domains

In [14]:
df['skill_domain'].value_counts()

skill_domain
Business Intelligence    11124
AI/ML/DL                  6591
Cloud & DevOps            2344
Data Engineering          1595
Data Science              1547
Name: count, dtype: int64

# ====================================
# Data Cleaning
# ====================================

## Step 1: Create a Working Copy

In [16]:
# Create a copy of the dataset
df_clean = df.copy()

## Step 2: Check Data Types Again

In [17]:
df_clean.dtypes

job_id                   int64
job_title               object
company_name            object
company_rating         float64
location                object
scraped_city            object
role_category           object
experience_raw          object
experience_min_yrs       int64
experience_max_yrs       int64
salary_raw              object
salary_min_lpa         float64
salary_max_lpa         float64
salary_disclosed          bool
skills_required         object
skills_count             int64
job_description         object
posted_date_raw         object
work_mode               object
company_size_bucket     object
job_url                 object
data_source             object
scraped_at              object
salary_tier             object
experience_tier         object
is_senior                 bool
primary_city            object
skill_domain            object
salary_midpoint_lpa    float64
days_since_posted        int64
is_fresher_friendly       bool
salary_negotiable         bool
dtype: o

## Step 3: Standardize Text Columns (removing unnecessary spaces)

In [18]:
text_columns = df_clean.select_dtypes(include='object').columns

for col in text_columns:
    df_clean[col] = df_clean[col].str.strip()

## Step 4: Verify Missing Values Again

In [19]:
df_clean.isnull().sum()

job_id                  0
job_title               0
company_name            0
company_rating          0
location                0
scraped_city            0
role_category           0
experience_raw          0
experience_min_yrs      0
experience_max_yrs      0
salary_raw              0
salary_min_lpa          0
salary_max_lpa          0
salary_disclosed        0
skills_required         0
skills_count            0
job_description         0
posted_date_raw         0
work_mode               0
company_size_bucket     0
job_url                53
data_source             0
scraped_at              0
salary_tier             0
experience_tier         0
is_senior               0
primary_city            0
skill_domain            0
salary_midpoint_lpa     0
days_since_posted       0
is_fresher_friendly     0
salary_negotiable       0
dtype: int64

## Step 5: Check Duplicate Job IDs

In [20]:
df_clean['job_id'].duplicated().sum()

0

## Step 6: Check Salary Columns

In [21]:
df_clean[['salary_min_lpa',
          'salary_max_lpa',
          'salary_midpoint_lpa']].describe()

,salary_min_lpa,salary_max_lpa,salary_midpoint_lpa
count,23201.000000,23201.000000,23201.000000
mean,1.383952,2.271790,1.827872
std,4.841570,7.469383,6.109373
min,0.000000,0.000000,0.000000
25%,0.000000,0.000000,0.000000
50%,0.000000,0.000000,0.000000
75%,0.000000,0.000000,0.000000
max,85.000000,90.000000,87.500000


# Step 6.1 Salary Data Investigation

# Objective:
# Determine whether salary values of 0 represent genuinely missing salary information.

# Count Jobs with Salary Information

In [22]:
salary_available = (df_clean['salary_midpoint_lpa'] > 0).sum()
salary_missing = (df_clean['salary_midpoint_lpa'] == 0).sum()

print(f"Salary Available : {salary_available}")
print(f"Salary Missing   : {salary_missing}")
print(f"Percentage Missing: {salary_missing/len(df_clean)*100:.2f}%")

Salary Available : 2766
Salary Missing   : 20435
Percentage Missing: 88.08%


# Check the salary_disclosed Flag

In [23]:
df_clean['salary_disclosed'].value_counts()

salary_disclosed
False    20433
True      2768
Name: count, dtype: int64

# Cross-check Salary vs Disclosure

In [24]:
pd.crosstab(
    df_clean['salary_disclosed'],
    df_clean['salary_midpoint_lpa'] == 0
)

salary_midpoint_lpa,False,True
salary_disclosed,,
False,0,20433
True,2766,2


### Salary Data Validation Summary

- 20,433 job postings did not disclose salary and therefore contain salary values of 0.
- 2,766 postings disclosed salary information with valid salary values.
- Only 2 records showed an inconsistency between the disclosure flag and the salary value.
- Salary values of 0 were retained because they represent "salary not disclosed" rather than missing or incorrect data.
- All salary-based analyses will use only records where `salary_disclosed = True`.

## Step 7: Experience Validation

In [25]:
df_clean[['experience_min_yrs',
          'experience_max_yrs']].describe()

,experience_min_yrs,experience_max_yrs
count,23201.000000,23201.000000
mean,4.135080,7.829059
std,2.933976,3.946834
min,0.000000,0.000000
25%,2.000000,5.000000
50%,4.000000,8.000000
75%,6.000000,10.000000
max,25.000000,31.000000


## Step 8: Company Rating Validation

In [26]:
df_clean['company_rating'].describe()

count    23201.000000
mean         3.586449
std          0.514250
min          1.000000
25%          3.400000
50%          3.600000
75%          3.800000
max          5.000000
Name: company_rating, dtype: float64

## Step 9: Save the Clean Dataset

In [27]:
df_clean.to_csv("cleaned_job_market.csv", index=False)

## Business Decision Log

### Decision 1
Salary values of 0 were retained because they indicate "salary not disclosed" rather than missing data. Salary-based analyses will therefore filter on `salary_disclosed = True`.

### Decision 2
Duplicate rows were not removed because no duplicate records were found.

### Decision 3
The original dataset was preserved, and all transformations were applied to `df_clean` to maintain reproducibility.

# =====================================================
# 6. Feature Engineering
# =====================================================

## In this section, we will add new columns that are necessary for us in future 

## Feature 1: Salary Range

In [28]:
import numpy as np

conditions = [
    (df_clean['salary_midpoint_lpa'] == 0),
    (df_clean['salary_midpoint_lpa'] < 5),
    (df_clean['salary_midpoint_lpa'] >= 5) & (df_clean['salary_midpoint_lpa'] < 10),
    (df_clean['salary_midpoint_lpa'] >= 10) & (df_clean['salary_midpoint_lpa'] < 20),
    (df_clean['salary_midpoint_lpa'] >= 20)
]

choices = [
    "Not Disclosed",
    "Low (Below 5 LPA)",
    "Medium (5–10 LPA)",
    "High (10–20 LPA)",
    "Premium (20+ LPA)"
]

df_clean['salary_range'] = np.select(
    conditions,
    choices,
    default="Unknown"
)

In [29]:
df_clean['salary_range'].value_counts()

salary_range
Not Disclosed        20435
High (10–20 LPA)      1071
Premium (20+ LPA)      837
Low (Below 5 LPA)      433
Medium (5–10 LPA)      425
Name: count, dtype: int64

## Business Question
## Which salary band has the highest number of job openings?
## Ans High (10-20 LPA)

## Feature 2: Experience Span

In [30]:
df_clean['experience_span'] = (
    df_clean['experience_max_yrs'] -
    df_clean['experience_min_yrs']
)

In [31]:
df_clean['experience_span'].describe()

count    23201.000000
mean         3.693979
std          1.902395
min        -12.000000
25%          2.000000
50%          4.000000
75%          5.000000
max         29.000000
Name: experience_span, dtype: float64

## Data Quality Investigation - Experience Span

## objective : Identify why some records have a negative experience span.

## Step 1: Find the Problematic Records

In [32]:
negative_exp = df_clean[df_clean['experience_span'] < 0]

negative_exp

,job_id,job_title,company_name,company_rating,location,scraped_city,role_category,experience_raw,experience_min_yrs,experience_max_yrs,...,experience_tier,is_senior,primary_city,skill_domain,salary_midpoint_lpa,days_since_posted,is_fresher_friendly,salary_negotiable,salary_range,experience_span
41,42,Walkin Drive | Bangalore | June 13 | Data Scie...,Tata Consultancy Services,3.3,Bengaluru,Bangalore,Data Scientist,13-Jun,13,8,...,Lead/Architect (9+ Yrs),False,Bangalore,AI/ML/DL,0.0,21,False,False,Not Disclosed,-5
45,46,Walk-in || Data Scientist,Tata Consultancy Services,3.3,Bengaluru,Bangalore,Data Scientist,13-Jun,13,8,...,Lead/Architect (9+ Yrs),False,Bangalore,AI/ML/DL,0.0,21,False,False,Not Disclosed,-5
48,49,Walk-in || Data Scientist(6+ yrs exp candidate...,Tata Consultancy Services,3.3,Bengaluru,Bangalore,Data Scientist,13-Jun,13,8,...,Lead/Architect (9+ Yrs),False,Bangalore,AI/ML/DL,0.0,21,False,False,Not Disclosed,-5
258,259,Walk-in || 13th Jun_ Data Scientist/Gen AI/MLO...,Ltimindtree,3.6,"Hybrid - Hyderabad, Bengaluru",Bangalore,Data Scientist,13-Jun,13,8,...,Lead/Architect (9+ Yrs),False,Hyderabad,AI/ML/DL,0.0,21,False,False,Not Disclosed,-5
690,691,Walk-in || Data Science Engineer,Creative Synergies,3.9,Hybrid - Bengaluru(Whitefield),Bangalore,Data Scientist,13-Jun,13,8,...,Lead/Architect (9+ Yrs),False,Bangalore,AI/ML/DL,0.0,21,False,False,Not Disclosed,-5
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
22417,22418,Walk in Interview Python Developer II Hyderabad,Tata Consultancy Services,3.3,"Mumbai (All Areas), Hyderabad, Pune",Mumbai,Python Developer,13-Jun,13,8,...,Lead/Architect (9+ Yrs),False,Mumbai,Data Engineering,0.0,21,False,False,Not Disclosed,-5
22426,22427,Walkin Drive -Python AI developer,Clover Infotech,3.4,Navi Mumbai,Mumbai,Python Developer,13-Jun,13,8,...,Lead/Architect (9+ Yrs),False,Mumbai,AI/ML/DL,10.0,21,False,False,High (10–20 LPA),-5
22631,22632,ArcGIS ESRI Developer,Capgemini,3.6,Navi Mumbai,Mumbai,Python Developer,20 Yrs,20,8,...,Lead/Architect (9+ Yrs),False,Mumbai,Data Science,0.0,14,False,False,Not Disclosed,-12
23032,23033,Walk-in || Dot Net Software Developer,Prudence Analytics Software Solutions,4.7,Mumbai (All Areas),Mumbai,Python Developer,13-Jun,13,8,...,Lead/Architect (9+ Yrs),False,Mumbai,Data Science,20.0,21,False,False,Premium (20+ LPA),-5


## Step 2: Check How Many Records Are Affected

In [33]:
negative_exp.shape

(90, 34)

In [34]:
negative_exp[
    [
        'job_id',
        'job_title',
        'experience_min_yrs',
        'experience_max_yrs',
        'experience_raw'
    ]
]

,job_id,job_title,experience_min_yrs,experience_max_yrs,experience_raw
41,42,Walkin Drive | Bangalore | June 13 | Data Scie...,13,8,13-Jun
45,46,Walk-in || Data Scientist,13,8,13-Jun
48,49,Walk-in || Data Scientist(6+ yrs exp candidate...,13,8,13-Jun
258,259,Walk-in || 13th Jun_ Data Scientist/Gen AI/MLO...,13,8,13-Jun
690,691,Walk-in || Data Science Engineer,13,8,13-Jun
...,...,...,...,...,...
22417,22418,Walk in Interview Python Developer II Hyderabad,13,8,13-Jun
22426,22427,Walkin Drive -Python AI developer,13,8,13-Jun
22631,22632,ArcGIS ESRI Developer,20,8,20 Yrs
23032,23033,Walk-in || Dot Net Software Developer,13,8,13-Jun


## Root Cause Analysis - Negative Experience Span

## Determine whether these 90 records are:

## - Scraping errors
## - Data entry errors
## - Legitimate values that need different handling

## Step 3: Display the Problematic Records

In [36]:
negative_exp[
    [
        'job_id',
        'job_title',
        'experience_raw',
        'experience_min_yrs',
        'experience_max_yrs'
    ]
].head(20)

,job_id,job_title,experience_raw,experience_min_yrs,experience_max_yrs
41,42,Walkin Drive | Bangalore | June 13 | Data Scie...,13-Jun,13,8
45,46,Walk-in || Data Scientist,13-Jun,13,8
48,49,Walk-in || Data Scientist(6+ yrs exp candidate...,13-Jun,13,8
258,259,Walk-in || 13th Jun_ Data Scientist/Gen AI/MLO...,13-Jun,13,8
690,691,Walk-in || Data Science Engineer,13-Jun,13,8
764,765,Walk-in || Big Data Engineer,13-Jun,13,8
1109,1110,Walk-in || Hiring For Snowflake Data Engineer,11-Jun,11,8
1672,1673,Walk-in || Data Scientist (AIML)_Walkin Drive_...,13-Jun,13,8
1701,1702,Walk-in || Data Scientist,13-Jun,13,8
2000,2001,Walk-in || Data Engineer,13-Jun,13,8


## Step 4: Check Unique Experience Formats

In [37]:
negative_exp['experience_raw'].value_counts()

experience_raw
13-Jun             77
20 Yrs              3
11-Jun              2
10-Jun              2
11 Jun - 10 Jul     2
20-Jun              1
17 May - 15 Aug     1
15 Yrs              1
19 May - 18 Jun     1
Name: count, dtype: int64

## Instead of removing thos 90 records, Flag these records as invalid and exclude them from analyses that depend on experience.

In [38]:
walkin_mask = df_clean['experience_span'] < 0
df_clean['experience_data_valid'] = ~walkin_mask

## preserve the original values instead of overwriting them.

In [39]:
df_clean['experience_issue'] = np.where(
    df_clean['experience_span'] < 0,
    'Walk-in Date Parsed as Experience',
    'Valid'
)

### Experience Data Validation

During data validation, 90 records (0.39% of the dataset) were identified with negative experience spans. Investigation showed these were primarily walk-in job postings where the web scraper incorrectly interpreted interview dates (e.g., "13-Jun") as experience values. Rather than deleting these records or fabricating corrected values, they were flagged using an `experience_data_valid` indicator and excluded from experience-based analyses.

## Feature 3: Experience Midpoint

In [40]:
df_clean['experience_midpoint'] = (
    df_clean['experience_min_yrs'] +
    df_clean['experience_max_yrs']
) / 2

In [41]:
df_clean[['experience_midpoint']].describe()

,experience_midpoint
count,23201.000000
mean,5.982070
std,3.344859
min,0.000000
25%,3.500000
50%,6.000000
75%,7.500000
max,26.000000


## Business Question
## What is the typical experience expected for a role?
## Ans Around 6 years of experience (Mean)

## Feature 4: Job Age Category

In [42]:
conditions = [
    df_clean['days_since_posted'] <= 7,
    (df_clean['days_since_posted'] > 7) & (df_clean['days_since_posted'] <= 30),
    df_clean['days_since_posted'] > 30
]

choices = [
    "Recent (0–7 Days)",
    "Active (8–30 Days)",
    "Older (30+ Days)"
]

df_clean['job_age_category'] = np.select(
    conditions,
    choices,
    default="Unknown"
)

In [43]:
df_clean['job_age_category'].value_counts()

job_age_category
Active (8–30 Days)    16600
Recent (0–7 Days)      6597
Older (30+ Days)          4
Name: count, dtype: int64

## Business Question
## Are most job postings recent or old?
## Ans Most of the jobs are active(8-30 days) followed by Recent jobs(0-7 days)

## Feature 5: Multi-Location Jobs

In [44]:
df_clean['is_multi_location'] = (
    df_clean['location']
    .str.contains(',', na=False)
)

In [53]:
df_clean['is_multi_location'].value_counts()

is_multi_location
False    16746
True      6455
Name: count, dtype: int64

## Business Question
## How many jobs are open across multiple locations?
## Ans 6455

## Feature 6: Company Rating Category

In [54]:
conditions = [
    df_clean['company_rating'] < 3,
    (df_clean['company_rating'] >= 3) & (df_clean['company_rating'] < 4),
    (df_clean['company_rating'] >= 4) & (df_clean['company_rating'] < 4.5),
    df_clean['company_rating'] >= 4.5
]

choices = [
    "Below Average",
    "Good",
    "Very Good",
    "Excellent"
]

df_clean['company_rating_category'] = np.select(
    conditions,
    choices,
    default="Not Rated"
)

In [55]:
df_clean['company_rating_category'].value_counts()

company_rating_category
Good             17001
Very Good         3122
Below Average     2021
Excellent         1057
Name: count, dtype: int64

## Business Question
## How are companies distributed by rating?
## Ans 17001 compamies are good(3-4), 3122 compamies are very good(4-4.5), 2021 companies are below avg.(less than 3) and 1057 companies are excellent(above 4.5)

# =============================================
# 7. Final Data Validation
# =============================================

# Step 1: Verify Dataset Shape

In [56]:
print(df_clean.shape)

(23201, 40)


# Step 2: Verify New Features

In [57]:
new_features = [
    'salary_range',
    'experience_span',
    'experience_midpoint',
    'job_age_category',
    'is_multi_location',
    'company_rating_category'
]

df_clean[new_features].head()

,salary_range,experience_span,experience_midpoint,job_age_category,is_multi_location,company_rating_category
0,Not Disclosed,3,8.5,Active (8–30 Days),False,Very Good
1,Not Disclosed,4,6.0,Active (8–30 Days),False,Very Good
2,Not Disclosed,5,6.5,Active (8–30 Days),True,Good
3,Not Disclosed,0,0.0,Active (8–30 Days),True,Good
4,Not Disclosed,5,7.5,Active (8–30 Days),True,Good


# Step 3: Check Missing Values Again

In [58]:
df_clean.isnull().sum().sort_values(ascending=False).head(15)

job_url                  53
job_title                 0
scraped_at                0
salary_tier               0
experience_tier           0
is_senior                 0
primary_city              0
skill_domain              0
salary_midpoint_lpa       0
days_since_posted         0
is_fresher_friendly       0
salary_negotiable         0
salary_range              0
experience_span           0
experience_data_valid     0
dtype: int64

# Step 4: Save Final Dataset

In [59]:
df_clean.to_csv(
    "job_market_analysis_final.csv",
    index=False
)

# ====================================================
# 8. Load Enhanced Dataset into MySQL
# ====================================================

# Step 1: Create Connection

In [65]:
from sqlalchemy import create_engine

engine = create_engine(
    "mysql+pymysql://root:27104720A@localhost:3306/job_market_analytics"
)

# Step 2: Load Data

In [66]:
df_clean.to_sql(
    name='jobs_final',
    con=engine,
    if_exists='replace',
    index=False
)

23201